### PR TITLE
Build using C++17 on CircleCI 

### PR DESCRIPTION
**Pull request by axic**
Part of #5913.

### CODE REVIEW

**Comment by axic** [comment]
@chriseth @christianparpart @ekpyron this can be merged to ensure our codebase remains C++17 compatible. It doesn't actually introduce any C++17 specific features.

File: .circleci/config.yml [hunk 1]
```diff
@@ -124,6 +127,30 @@ jobs:
           paths:
             - "*"
 
+  build_x86_linux_cxx17:
+    docker:
+      - image: buildpack-deps:bionic
```
Comment (ekpyron) [inline .circleci/config.yml]: If this build is merely meant to check for forwards compatibility, we could probably change this to ``buildpack-deps:disco``, which should contain cvc4 1.6 - that way we could keep cvc4 enabled below.

Comment (axic) [reply .circleci/config.yml]: Can you add a commit testing that assumption? I'm fine with that, should enable us to ensure C++17 compatibility in the SMT subsystem.

Comment (ekpyron) [reply .circleci/config.yml]: Let's see what happens!

Comment (ekpyron) [reply .circleci/config.yml]: Looks like it works (SMT stuff just compiled on the CI run).

Comment (axic) [reply .circleci/config.yml]: You want to approve the PR then? 😉

**Review by ekpyron — DISMISSED** [review]

File: .circleci/config.yml [hunk 2]
```diff
@@ -124,6 +127,29 @@ jobs:
...
+            apt-get -qq update
+            apt-get -qy install cmake libboost-regex-dev libboost-filesystem-dev libboost-test-dev libboost-system-dev libboost-program-options-dev libcvc4-dev
+            ./scripts/install_obsolete_jsoncpp_1_7_4.sh
+      - run: *setup_prerelease_commit_hash
+      - run: *run_build
+      - store_artifacts: *solc_artifact
+      - persist_to_workspace:
+          root: build
+          paths:
+            - "*"
```
Comment (chriseth) [inline .circleci/config.yml]: Why is this needed?

Comment (axic) [reply .circleci/config.yml]: The block was copied from `build_x86_linux`. Even `build_x86_archlinux` saves some artifacts. Do we really need to save artifacts for these targets not used for testing?

Comment (ekpyron) [reply .circleci/config.yml]: For archlinux there's ``test_x86_archlinux`` running the non-ipc tests, which needs some artifacts. And the artifacts of ``build_x86_linux`` are used in ``test_x86_linux``, but I guess we don't need to store any artifacts here, if we don't want to test the C++17 build.

Comment (chriseth) [reply .circleci/config.yml]: Some rules only store the binaries, which is sufficient. Other rules store the whole build directory because that is needed if you want to calculate coverage.

Comment (chriseth) [inline .circleci/config.yml]: Please remove.

**Review by chriseth — DISMISSED** [review]
Remove

**Comment by axic** [comment]
@chriseth @ekpyron should be ready now

**Review by ekpyron — APPROVED** [review]
